Automattic/woocommerce-payments
# Allow the use of saved cards when UPE is active.

Fixes #1870 

#### Changes proposed in this Pull Request

Allow the use of saved cards when UPE is active:
Paying with a saved card does not use the flow of UPE, in this PR, when a token is selected on checkout, we will use the method from WCPay Payment Gateway to process the payment, instead of the UPE flow.

Checking out and saving a new payment method was implemented [on this PR](https://github.com/Automattic/woocommerce-payments/pull/1919).

It also fixes a bug that was preventing checking out with other payment methods due to a nonce verification.
This PR changes the order of the checks on function `process_redirect_payment`. 
The first check will verify that it is UPE Payment gateway and return if not.

Managing saved payment methods is out of scope for this PR.









#### Testing instructions




Checking out with Cards saved before UPE:
* Please go to Payments > Settings and ensure that WooCommerce Payments is enabled.
* With UPE disabled, save a credit card from My Account > Payment Methods
* Enter your WC Options table and add the feature flag _wcpay_feature_upe with a value of 1. Stripe UPE payment gateway should now replace the Credit Card and be enabled at checkout.
* Add an item from your shop to your cart and proceed through a regular checkout. At checkout, select "WooCommerce Payments" and you should see the UPE appear below the payment gateway.
* Verify that you can checkout with previously saved payment method
---
Checking out with Cards saved after UPE:
* Add an item from your shop to your cart and proceed again through a regular checkout. 
* Enter the credit card details, check the box to save payment method and checkout.
* Add an item from your shop to your cart and proceed again through a regular checkout. 
* Check that the payment method was saved.
* Verify that you can checkout with previously saved payment method

-------------------
